src/Database/LSMTree/Internal.hs

```diff
@@ -1278,3 +1302,140 @@ duplicate t@Table{..} = do
+
+            t <-
+              newWith
+                reg
+                sesh
+                seshEnv
+                conf
+                (error "unions: ArenaManager") -- TODO
+                content
+
+            pure (seshState, t)
+  where
+    n = V.length ts
+
+-- | Like 'vmatchBy', but the match function is @(==)@.
+vmatch :: Eq a => V.Vector a -> Either (Int, Int) a
+vmatch = vmatchBy (==)
+
+-- | Check that all values in the vector match. If so, return the matched value.
+-- If there is a mismatch, return the vector indices of the mismatching values.
+--
+-- Assumes the input vector is non-empty.
+vmatchBy :: (a -> a -> Bool) -> V.Vector a -> Either (Int, Int) a
+vmatchBy cmp xs0 =
+    case V.uncons xs0 of
+      Nothing ->
+        error "vmatch: empty vector "
+      Just (x, xs) ->
+        case V.iforM_ xs $ vmatchOne x of
+          Left i   -> Left (0, i)
+          Right () -> Right x
+  where
+    vmatchOne x i y =
+      if (x `cmp` y)
+        then Left i
+        else Right ()
+
+-- | Check that all tables in the session match. If so, return the matched
+-- session. If there is a mismatch, return the vector indices of the mismatching
+-- tables.
+--
+-- Assumes the input vector is non-empty.
+--
+-- TODO: compare LockFileHandle instead of SessionRoot (?). We can write an Eq
+-- instance for LockFileHandle based on pointer equality, just like base does
+-- for Handle.
+vmatchSessions ::
+     (MonadSTM m, MonadThrow m)
+  => V.Vector (Table m h)
+  -> m (Either (Int, Int) (Session m h))
+vmatchSessions ts0
+  | Just (t, ts) <- V.uncons ts0
+  = do
+      root <- getSessionRoot t
+      eith <- go root 1 ts
+      pure $ case eith of
+        Left i   -> Left (0, i)
+        Right () -> Right (tableSession t)
+  | otherwise
+  = error "vmatchSessions: empty vector"
+  where
+    -- Check that the session roots for all tables are the same. There can only
+    -- be one *open/active* session per directory because of cooperative file
+    -- locks, so each unique *open* session has a unique session root.
+    --
+    -- In remote cases, the session can be be closed, after which a new session
+    -- can be opened in the same root directory. In this case, the session root
+    -- is technically not unique. This does not mean that tables in @ts0@ could
+    -- have different sessions with the same root, however. You could prove this
+    -- by contradiction, assuming that there are two tables in @ts0@ that have
+    -- different open sessions with the same session root.
+    go root !i ts
+      | Just (t', ts') <- V.uncons ts
+      = do
+          root' <- getSessionRoot t'
+          if root == root'
+            then go root (i+1) ts'
+            else pure (Left i)
+      | otherwise
+      = pure (Right ())
+
+    -- We do not have to keep the session open for longer than it takes to read
+    -- the session root.
+    getSessionRoot t =  withOpenSession (tableSession t) $ pure . sessionRoot
```

Comment:
So on reflection, I think this is wrong, but fixing it also makes explaining why it's correct easy.

I suggest this as a comment:

> Open sessions hold an exclusive lock on a file in the session directory. Two different open sessions cannot hold the same exclusive lock at the same time. Therefore to check if two tables share the same session, we just need to hold open both table's sessions and check that they are using the same session directory. If it's the same directory, it's the same session.

So we _must_ hold open all the sessions while we do the comparison of the session roots. So the `withOpenSession` must scope over the comparison of the roots, not be used per-session to extract the roots.

I was wondering if we need to canonicalise the session root dir when we open it, but I think we do not. Any second session open with a different path that resolves to the same directory will also result in trying to take the same exclusive lock, which will fail.

Yes it might be slightly clearer to argue if one compares the open lock file handles themselves, but that's non-trivial to do and depends on the lock implementation. So session root dirs is sufficient. But the sessions must be held open during the comparison for it to be valid.

Reply:
Your suggestion makes the correctness argument simpler, but I'd disagree that we must scope `withOpenSession` over the comparison to make it correct. How are you going to construct two tables that are passed to `unions` which have the same root but are in different open sessions? 

Say I have tables `t1` and `t2`. I read `getSessionRoot t1` succesfully, and I get out some directory `dir`. I then read `getSessionRoot t2` succesfully and get out `dir` again, but somehow it's not the same session. It would mean that between the reads, `t1`'s session was closed, and `t2` was assigned a new open session in `dir`. This isn't possible: `t2`'s session has to have been opened before we can open table `t2`. `t2` can't be created without a session and assigned an open session later

Reply:
I suppose the argument is that the tables being open implies their parent sessions are also open.

In your argument with `t1` and `t2`, don't you need to verify that `t1` and `t2` are open simultaneously (or were at some point open simultaneously)? Otherwise if e.g. `t1` is already closed then `t2` can have come from a different session opened in the same directory.

Reply:
Ok, so we have tables t0..tn. Each of these must have been open (or we would not have a table handle). Each one can only monotonically go from open to closed. So if `getSessionRoot` itself checks that the table is open (and thus session is open), then sequentially checking each one is open (and subsequently checking they have the same root) does guarantee that they were at some point all open simultaneously. So yeah ok.